Migrate thnn_conv_depthwise2d from THC to ATen

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #62006

Closes gh-24646, gh-24647

There is no `TensorIterator` equivalent to these kernels so this is just
migrating the existing kernels over to the ATen style.

I've benchmarked for contiguous tensors with this script:
```
import torch
shape = (10, 10, 100, 100)
x = torch.randn(*shape, device='cuda')
w = torch.randn((10, 1, 5, 5), device='cuda')

for _ in range(100):
    torch.nn.functional.conv2d(x, w, groups=10)
```

and similarly for backwards. I see these as the same to within measurement error.

|                   | Master Forward (us) | This PR Forward (us) |
|------------------:|:-------------------:|:--------------------:|
|           Forward |        133.5        |         133.6        |
|  Backward (input) |        1,102        |         1,119        |
| Backward (weight) |        2,220        |         2,217        |

Differential Revision: [D29883676](https://our.internmc.facebook.com/intern/diff/D29883676)

peterbell10 committed:
Migrate thnn_conv_depthwise2d from THC to ATen

Closes gh-24646, gh-24647

There is no `TensorIterator` equivalent to these kernels so this is just
migrating the existing kernels over to the ATen style.

I've benchmarked for contiguous tensors with this script:
```
import torch
shape = (10, 10, 100, 100)
x = torch.randn(*shape, device='cuda')
w = torch.randn((10, 1, 5, 5), device='cuda')

for _ in range(100):
    torch.nn.functional.conv2d(x, w, groups=10)
```

and similarly for backwards. I see these as the same to within measurement error.

|                   | Master Forward (us) | This PR Forward (us) |
|------------------:|:-------------------:|:--------------------:|
|           Forward |        133.5        |         133.6        |
|  Backward (input) |        1,102        |         1,119        |
| Backward (weight) |        2,220        |         2,217        |

[ghstack-poisoned]

peterbell10 committed:
Update on "Migrate thnn_conv_depthwise2d from THC to ATen"


Closes gh-24646, gh-24647

There is no `TensorIterator` equivalent to these kernels so this is just
migrating the existing kernels over to the ATen style.

I've benchmarked for contiguous tensors with this script:
```
import torch
shape = (10, 10, 100, 100)
x = torch.randn(*shape, device='cuda')
w = torch.randn((10, 1, 5, 5), device='cuda')

for _ in range(100):
    torch.nn.functional.conv2d(x, w, groups=10)
```

and similarly for backwards. I see these as the same to within measurement error.

|                   | Master Forward (us) | This PR Forward (us) |
|------------------:|:-------------------:|:--------------------:|
|           Forward |        133.5        |         133.6        |
|  Backward (input) |        1,102        |         1,119        |
| Backward (weight) |        2,220        |         2,217        |

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/cuda/DepthwiseConv2d.cu:
```diff
@@ -1,29 +1,597 @@
+
+template <typename scalar_t, int ndim, template <typename U> class PtrTraits = DefaultPtrTraits>
+PackedTensorAccessor32<scalar_t, ndim, PtrTraits> dummy_packed_accessor32() {
+  std::array<int64_t, ndim> zeros;
+  zeros.fill(0);
```
```suggestion
  std::array<int64_t, ndim> zeros{};
```

ngimel commented on aten/src/ATen/native/cuda/DepthwiseConv2d.cu:
```diff
@@ -1,29 +1,597 @@
+  int64_t batchSize = in_sizes[0];
+  int64_t height = in_sizes[2];
+  int64_t width = in_sizes[3];
+  int64_t outputHeight = (height + 2 * padH - (dilationH * (kH - 1) + 1)) / dH + 1;
```
nit: you can use `conv_output_size` here (to make sure it doesn't diverge with how other convolution implementations define it)

ngimel commented on aten/src/ATen/native/cuda/DepthwiseConv2d.cu:
```diff
@@ -1,29 +1,597 @@
-    const Tensor & weight,
+  // One thread per output value
+  TORCH_CHECK(canUse32BitIndexMath(input) && canUse32BitIndexMath(output));
+  unsigned n = output.numel();
```
nit: if canUse32BitIndexMath in the line above is true, `n` can be `int32_t` (signed types are marginally more efficient)

ngimel commented on aten/src/ATen/native/cuda/DepthwiseConv2d.cu:
```diff
@@ -1,29 +1,597 @@
+                  const Tensor &output,
+                  const Tensor &weight,
+                  const Tensor &bias,
+                  int kW, int kH,
```
nit: all these should be const, or they can be replaced with `params`

ngimel commented on aten/src/ATen/native/cuda/DepthwiseConv2d.cu:
```diff
@@ -1,29 +1,597 @@
+    auto weight = weight_.expect_contiguous();
+    conv_depthwise2d_backward_out(
+        *self, *grad_output, grad_input, *weight,
+        kernel_size[0], kernel_size[1],
```
should it be `kernel_size[1], kernel_size[0]` here, like in the forward call? Arguments are `int kW, int kH,`. This legacy hw flip is so confusing

peterbell10 committed:
Update on "Migrate thnn_conv_depthwise2d from THC to ATen"


Closes gh-24646, gh-24647

There is no `TensorIterator` equivalent to these kernels so this is just
migrating the existing kernels over to the ATen style.

I've benchmarked for contiguous tensors with this script:
```
import torch
shape = (10, 10, 100, 100)
x = torch.randn(*shape, device='cuda')
w = torch.randn((10, 1, 5, 5), device='cuda')

for _ in range(100):
    torch.nn.functional.conv2d(x, w, groups=10)
```

and similarly for backwards. I see these as the same to within measurement error.

|                   | Master Forward (us) | This PR Forward (us) |
|------------------:|:-------------------:|:--------------------:|
|           Forward |        133.5        |         133.6        |
|  Backward (input) |        1,102        |         1,119        |
| Backward (weight) |        2,220        |         2,217        |

[ghstack-poisoned]

peterbell10 commented on aten/src/ATen/native/cuda/DepthwiseConv2d.cu:
```diff
@@ -1,29 +1,597 @@
+  // accumulate prior to writing the global value
+  extern __shared__ char smem[];
+  acc_t* buf = reinterpret_cast<acc_t*>(smem);
+  acc_t tval = reduceBlock(buf, blockDim.x, grad,
```
Thanks, I didn't know about `BlockReduceSum`.

ngimel commented on aten/src/ATen/native/cuda/DepthwiseConv2d.cu:
```diff
@@ -1,29 +1,596 @@
+    const auto input_a = input.packed_accessor32<scalar_t, 4>();
+    const auto grad_weight_a = grad_weight.packed_accessor32<scalar_t, 4>();
+    using acc_t = at::acc_type<scalar_t, true>;
+    int smem = ((block.x + C10_WARP_SIZE - 1) / C10_WARP_SIZE) * sizeof(acc_t);
```
I think blockReduceSum cannot handle non-multiple-of-warp number of threads, so this can be changed back to `int smem = block.x * sizeof(acc_t)`, and we can add TORCH_INTERNAL_ASSERT that (block.x % C10_WARP_SIZE ==0)

ngimel commented on aten/src/ATen/native/cuda/DepthwiseConv2d.cu:
```diff
@@ -1,29 +1,596 @@
+    conv_depthwise2d_backward_out(
+        *self, *grad_output, grad_input, *weight,
+        kernel_size[1], kernel_size[0],
+        stride[1], stride[0],
```
this is scary, can we add a test that would have caught it?

peterbell10 replied:
I've added a `gradcheck` test that fails for the previous version.

peterbell10 commented on aten/src/ATen/native/ConvUtils.h:
```diff
@@ -83,12 +84,12 @@ static inline Tensor reshape_bias(int64_t dim, const Tensor& bias) {
 static inline bool cudnn_conv_use_channels_last(const at::Tensor& input, const at::Tensor& weight) {
   // disable NHWC for float64 input.
-  if (!detail::getCUDAHooks().compiledWithCuDNN() ||
+  if (!at::detail::getCUDAHooks().compiledWithCuDNN() ||
```
Without this I get compilation errors because it's looking in the wrong namespace:
```
/home/peter/git/pytorch/aten/src/ATen/native/ConvUtils.h(87): error: namespace "at::native::detail" has no member "getCUDAHooks"
```

peterbell10 committed:
Update on "Migrate thnn_conv_depthwise2d from THC to ATen"


Closes gh-24646, gh-24647

There is no `TensorIterator` equivalent to these kernels so this is just
migrating the existing kernels over to the ATen style.

I've benchmarked for contiguous tensors with this script:
```
import torch
shape = (10, 10, 100, 100)
x = torch.randn(*shape, device='cuda')
w = torch.randn((10, 1, 5, 5), device='cuda')

for _ in range(100):
    torch.nn.functional.conv2d(x, w, groups=10)
```

and similarly for backwards. I see these as the same to within measurement error.

|                   | Master Forward (us) | This PR Forward (us) |
|------------------:|:-------------------:|:--------------------:|
|           Forward |        133.5        |         133.6        |
|  Backward (input) |        1,102        |         1,119        |
| Backward (weight) |        2,220        |         2,217        |

[ghstack-poisoned]

ngimel commented:
Test failures looks unrelated.

ngimel commented:
@ngimel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29883676).

peterbell10 commented:
@ngimel `check_backward_compatibility` is failing and I missed it because that CI job wasn't run on this PR for some reason.

ejguan commented:
Reverting this PR since it breaks `check_backward_compatibility`. Please add BC-breaking changes to the allow list  in https://github.com/pytorch/pytorch/blob/master/test/backward_compatibility/check_backward_compatibility.py
